rfcs/0014-wal_on_disk_format.md

```diff
@@ -0,0 +1,293 @@
+| number of records N (4-bytes unsigned integer, little endian)  |
++----------------------------------------------------------------+
+| version of format (2-bytes, unsigned integer, little endian)   |
++----------------------------------------------------------------+
+```
+
+Ideally, the in-memory data structure for the WAL stores the incoming records in ingestion order, so that the records do not need
+to be re-ordered before the flush. A simple FIFO data structure like a queue is recommended.
+
+<!-- A detailed description of the proposed change. Include diagrams, examples, schemas, and pseudo-code as appropriate. -->
+
+<!-- TOC --><a name="impact-analysis"></a>
+## Impact Analysis
+
+SlateDB features and components that this RFC interacts with. Check all that apply.
+
+<!-- TOC --><a name="core-api-query-semantics"></a>
+### Core API & Query Semantics
+
+- [ ] Basic KV API (`get`/`put`/`delete`)
+- [ ] Range queries, iterators, seek semantics
+- [ ] Range deletions
+- [ ] Error model, API errors
+
+<!-- TOC --><a name="consistency-isolation-and-multi-versioning"></a>
+### Consistency, Isolation, and Multi-Versioning
+
+- [ ] Transactions
+- [ ] Snapshots
+- [ ] Sequence numbers
+
+<!-- TOC --><a name="time-retention-and-derived-state"></a>
+### Time, Retention, and Derived State
+
+- [ ] Logical clocks
+- [ ] Time to live (TTL)
+- [ ] Compaction filters
+- [ ] Merge operator
+- [ ] Change Data Capture (CDC)
+
+<!-- TOC --><a name="metadata-coordination-and-lifecycles"></a>
+### Metadata, Coordination, and Lifecycles
+
+- [ ] Manifest format
+- [ ] Checkpoints
+- [ ] Clones
+- [ ] Garbage collection
+- [ ] Database splitting and merging
+- [ ] Multi-writer
+
+<!-- TOC --><a name="compaction"></a>
+### Compaction
+
+- [ ] Compaction state persistence
+- [ ] Compaction filters
+- [ ] Compaction strategies
+- [ ] Distributed compaction
+- [ ] Compactions format
+
+<!-- TOC --><a name="storage-engine-internals"></a>
+### Storage Engine Internals
+
+- [X] Write-ahead log (WAL)
+- [ ] Block cache
+- [ ] Object store cache
+- [ ] Indexing (bloom filters, metadata)
+- [ ] SST format or block format
+
+<!-- TOC --><a name="ecosystem-operations"></a>
+### Ecosystem & Operations
+
+- [ ] CLI tools
+- [ ] Language bindings (Go/Python/etc)
+- [ ] Observability (metrics/logging/tracing)
+
+<!-- TOC --><a name="operations"></a>
+## Operations
+
+<!-- TOC --><a name="performance-cost"></a>
+### Performance & Cost
+
+<!-- Describe performance and cost implications of this change. -->
+
+- **Latency (reads/writes/compactions):** 
+The latency for writes with durability guarantee mainly depends on the configured size of the encoded WAL object 
+(see `max_wal_bytes_size`) that is used to decide when the in-memory WAL data structure is flushed to object storage.
+Since the new persistence format can batch more records than the currently used SST format given the size of the 
+encoded WAL object, the latency might be higher.
```

Comment:
not sure I follow this. We still have a flush interval right? Which should let you control the latency when it's configured?